Sources/NIOTestUtils/ByteToMessageDecoderVerifier.swift

```diff
@@ -0,0 +1,211 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+public enum ByteToMessageDecoderVerifier {
+    /// - seealso: verifyDecoder(inputOutputPairs:decoderFactory:)
+    ///
+    /// Verify `ByteToMessageDecoder`s with `String` inputs
+    public static func verifyDecoder<Decoder: ByteToMessageDecoder>(stringInputOutputPairs: [(String, [Decoder.InboundOut])],
+                                                                    decoderFactory: @escaping () -> Decoder) throws where Decoder.InboundOut: Equatable {
+        let alloc = ByteBufferAllocator()
+        let ioPairs = stringInputOutputPairs.map { (ioPair: (String, [Decoder.InboundOut])) -> (ByteBuffer, [Decoder.InboundOut]) in
+            var buffer = alloc.buffer(capacity: ioPair.0.utf8.count)
+            buffer.writeString(ioPair.0)
+            return (buffer, ioPair.1)
+        }
+        return try ByteToMessageDecoderVerifier.verifyDecoder(inputOutputPairs: ioPairs, decoderFactory: decoderFactory)
+    }
+
+    /// Verifies a `ByteToMessageDecoder` by performing a number of tests.
+    ///
+    /// This method is mostly useful in unit tests for `ByteToMessageDecoder`s. It feeds the inputs from
+    /// `inputOutputPairs` into the decoder in various ways and expects the decoder to produce the outputs from
+    /// `inputOutputPairs`.
+    ///
+    /// The verification is performing various tests, for example:
+    ///
+    ///  - drip feeding the bytes, one by one
+    ///  - sending many messages in one `ByteBuffer`
+    ///  - sending each complete message in one `ByteBuffer`
+    ///
+    /// For `ExampleDecoder` that produces `ExampleDecoderOutput`s you would use this method the following way:
+    ///
+    ///     var exampleInput1 = channel.allocator.buffer(capacity: 16)
+    ///     exampleInput1.writeString("example-in1")
+    ///     var exampleInput2 = channel.allocator.buffer(capacity: 16)
+    ///     exampleInput2.writeString("example-in2")
+    ///     let expectedInOuts = [(exampleInput1, ExampleDecoderOutput("1")),
+    ///                           (exampleInput2, ExampleDecoderOutput("2"))
+    ///                          ]
+    ///     XCTAssertNoThrow(try ByteToMessageDecoderVerifier.verifyDecoder(inputOutputPairs: expectedInOuts,
+    ///                                                                     decoderFactory: { ExampleDecoder() }))
+    public static func verifyDecoder<Decoder: ByteToMessageDecoder>(inputOutputPairs: [(ByteBuffer, [Decoder.InboundOut])],
+                                                                    decoderFactory: @escaping () -> Decoder) throws where Decoder.InboundOut: Equatable {
+        typealias Out = Decoder.InboundOut
```

Comment:
Nit: this `typealias` is used inconsistently in this function: we may as well use it everywhere, no?